src/util/artifacts.js

```diff
@@ -18,8 +23,20 @@ export async function prepareArtifactsDir(
     }
   } catch (error) {
     if (isErrorWithCode('ENOENT', error)) {
-      log.debug(`Creating artifacts directory: ${artifactsDir}`);
-      await fs.mkdir(artifactsDir);
+      try {
```

Comment:
I tested this function a bit locally, especially to:
- verify that the main scenarios are covered
- that the error message raised is nice and clear
- that all these try/catch blocks are needed

I think that it is pretty near to what we want it to be, but I'd like to propose some further tweaks on it, e.g.:
- tweak the error message to be more readable
- reorder the `if/if else/else` inside the `try/catch`, so that the `EACCES` raised from `fs.stat` is the first one, then the `ENOENT` (also raised from fs.stat) and finally `EACCES` error that can be raised from the `fs.mkdir`
- put an inline comment in this `isErrorWithCode` if branches (so that we know which scenario anyone of them is going to handle, e.g.

```
  try {
    const stats = await fs.stat(artifactsDir);
    if (!stats.isDirectory()) {
      throw new UsageError(
        `"${artifactsDir}" exists and it is not a directory.`);
    }
  } catch (error) {
    if (isErrorWithCode('EACCES', error)) {
      // Handle errors when the artifactsDir cannot be accessed.
      throw new UsageError(
        `"Cannot access ${artifactsDir}", user lacks permissions.`);
    } else if (isErrorWithCode('ENOENT', error)) {
      // Handle errors when the artifactsDir doesn't exists yet.
      try {
        log.debug(`Creating artifacts directory: ${artifactsDir}`);
        await fs.mkdir(artifactsDir);
      } catch (mkdirErr) {
        if (isErrorWithCode('EACCES', mkdirErr)) {
          // Handle errors when the artifactsDir cannot be created for lack of permissions.
          throw new UsageError(
            `Cannot create "${artifactsDir}", user lacks permissions.`);
        } else {
          throw mkdirErr;
        }
      }
    } else {
      throw error;
    }
  }
```

also, it would be nice to use something like [node-mkdirp](https://github.com/substack/node-mkdirp) instead of `fs.mkdir` so that we create the parent dirs if they don't exists yet (NOTE: `mkdirp` uses the usual nodejs callback style, and so we need to wrap it into a promise to be able to use it with `await`)

Finally, if I'm not wrong, `fs.stat` is going to raise an error if we cannot access the artifacts dir, but if the artifacts dir already exists and we cannot create any file in it (for lack of permissions), stat is not going to complain, as a workaround we can probably use `fs.access` to check that we can actually create a file in it, e.g. something like:

```
  try {
    const stats = await fs.stat(artifactsDir);
    if (!stats.isDirectory()) {
      throw new UsageError(
        `"${artifactsDir}" exists and it is not a directory.`);
    }

    // Check that we can write in the artifacts dir.
    try {
      await fs.access(path.join(artifactsDir, "someRandomFileNameHere"), fs.W_OK);
    } catch (error) {
      if (isErrorWithCode('EACCES', error)) {
        throw new UsageError(
          `"${artifactsDir}" exists but the user lacks permissions on it.`);
      }
    }
  } catch (error) {
```